### DISCUSSION TITLE
How to define external Vim functions?

### DISCUSSION THREAD

**Discussion by jphalip** [JetBrains/ideavim#1041] [Q&A] [answered]
Hi,

I'd like to create some custom Vim functions. It looks like all the standard functions in IdeaVim are annotated with `@VimscriptFunction`, but that annotation comes from the `annotation-processors` kotlin project, which I'm not sure is accessible from outside of IdeaVim.

Could you please clarify it it's in fact possible to define your own functions, and if so, how?

Thanks!

Julien

**Comment by citizenmatt** [comment, marked as answer]
If you're writing an extension to IdeaVim, you'll most likely implement `VimExtension` to register mappings, etc. You can also export functions here, using the `VimExtensionFacade.exportScriptFunction` method.

Here's the method signature:

https://github.com/JetBrains/ideavim/blob/791edbd29b6de7db64fd622744d21001a8c54d17/src/main/java/com/maddyhome/idea/vim/extension/VimExtensionFacade.kt#L221-L229

This essentially corresponds to the `:function` command and defines the function, with name, arguments, default values, etc. You can also specify the `range`, `abort`, `dict` and `closure` flags (see `:help func-range`, `:help func-abort`, etc.). Finally, you need to pass an instance of the `ScriptFunction` interface:

https://github.com/JetBrains/ideavim/blob/791edbd29b6de7db64fd622744d21001a8c54d17/src/main/java/com/maddyhome/idea/vim/extension/VimExtensionFacade.kt#L278-L280

This is easily implemented, takes in an editor, the current context and a map of argument to `VimDataType` values. You don't get quite as much flexibility as a normal script function, especially with additional optional arguments, but you can most likely implement what you need. The return value should be an instance of `ExecutionResult`, which can be success, error, break, continue or an instance with a return value.

I'm only aware of one usage of this method, and that's to support custom operators from extensions with the `'operatorfunc'` option. Some of the built in extensions will export a script function and create bindings that `set opfunc=CommentaryOperatorFunc` and then invoke the `g@` command, just like they would do as a standard Vim script plugin.

They use the `VimExtensionFacade.exportOperatorFunction` helper to export the function, and it's a good example of how to export custom code as a script function:

https://github.com/JetBrains/ideavim/blob/791edbd29b6de7db64fd622744d21001a8c54d17/src/main/java/com/maddyhome/idea/vim/extension/VimExtensionFacade.kt#L257-L276

It registers the script function and provides an (implicit) implementation of the `ScriptFunction` interface that converts the `'operatorfunc'` arguments into parameters for the `OperatorFunction.apply` method.

**Comment by jphalip** [comment]
Thanks @citizenmatt for your thorough reply. I'll look into this and come back if I run into any issues.
